Implemented ICU dependency

DenuxPlays opened the pull request:
This is a PR to test the implementation of ICU.

Ref: https://github.com/keycloak/keycloak/issues/34962#issuecomment-2597603323

I think that the minimum required windows version is: `Version 1703`
So this would include Windows 10 and Windows Server 2016
Because since then icu is included in Windows.
BUT it should be possible install icu even before this update.

On Linux icu4j requries the icu package to be installed.
But it is already preinstalled by a lot of distros (tested it on fedora 41).

DenuxPlays committed:
Implemented ICU dependency

ahus1 commented on services/pom.xml:
```diff
@@ -50,6 +50,11 @@
             <artifactId>${keycloak.crypto.artifactId}</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.ibm.icu</groupId>
+            <artifactId>icu4j</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
Can you please elaborate why you choose provided here?  With this setting, the JAR would not be packaged.

ahus1 commented on services/src/main/java/org/keycloak/theme/beans/LocaleBean.java:
```diff
@@ -17,23 +17,22 @@
 
-    private static final Set<String> RTL_LANGUAGE_CODES =
-            Set.of("ar", "dv", "fa", "ha", "he", "iw", "ji", "ps", "sd", "ug", "ur", "yi");
+    private static final HashMap<String, Boolean> CACHED_RTL_LANGUAGE_CODES = new HashMap<>();
```
This needs to be at least a `ConcurrentHashMap` as multiple threads might add values concurrently.

ahus1 commented on services/src/main/java/org/keycloak/theme/beans/LocaleBean.java:
```diff
@@ -73,6 +72,18 @@ public boolean isRtl() {
     }
 
+    public boolean isRtl(String languageTag) {
+        if (CACHED_RTL_LANGUAGE_CODES.containsKey(languageTag)) {
+            return CACHED_RTL_LANGUAGE_CODES.get(languageTag);
+        }
+
+        boolean isRTL = new ULocale(languageTag).isRightToLeft();
+
+        CACHED_RTL_LANGUAGE_CODES.put(languageTag, isRTL);
+
+        return isRTL;
```
Try to leverage modern APIs like `HashMap#computeIfAbsent` here to reduce boilerplate code and avoid concurrency issues.

ahus1 reviewed (commented):
Thank you for the PR. Some questions below. 
Please provide at least a unit test.
Also, add a note to the docs which versions of Windows you think support this, or if additional dependencies are used by ICU.

DenuxPlays commented:
How do I unit test that?
Especially how do I get an instance of RealmModel?

You don't have any mocking tools

DenuxPlays committed:
addressed review

DenuxPlays commented:
Adressed your review.
Also I've added a unit test and updated the pr description.
Unfortunately I couldn't get the unit test to work (even on the main branch).

ahus1 reviewed (changes requested):
I don't think this is a good showcase for ICU. 

The same can be achieved using https://github.com/DenuxPlays/keycloak/pull/1

If you agree with the approach, we can proceed without ICU and get it merged.

ahus1 commented:
Reading more on the previous discussion, I see you brought it up earlier. While there are some exception, I would be ok to approve it given the licensing issues with ICU. 

See https://github.com/keycloak/keycloak/issues/33833#issuecomment-2446965307 for the exceptions.

DenuxPlays commented:
> I don't think this is a good showcase for ICU. 
> 
> 
> 
> The same can be achieved using https://github.com/DenuxPlays/keycloak/pull/1
> 
> 
> 
> If you agree with the approach, we can proceed without ICU and get it merged. 

No it cannot 
See here: https://github.com/keycloak/keycloak/issues/33833#issuecomment-2446965307

ahus1 commented:
Feels like a chat :-D we are commenting concurrently. 

Still while there are exceptions, it still seems to be better than adding a dependency like ICU where we need to get a license exception first.

DenuxPlays commented:
Hahaha yes

I know that just for this one it would be too little of an advantage.
But combined with the other ticket (and probably more) it would be worth.

I know that this isnt in the pr currently.

They are also more things that icu can do like ordinality.
Like 1st, 2nd etc because this is way more complex in other languages.
This can also be handled by icu

ahus1 commented:
See my other comments for example in #33833. I don't see the benefit of ICU as of now when I compare it to the alternatives. So please go ahead with an approach that leverages Bidi for RTL.
